lib/chef/run_list/run_list_expansion.rb

```diff
@@ -156,8 +169,26 @@ def expand_run_list_items(items, included_by="top level")
         end
       end
 
+      # Recursive helper to decode the non-nested hash form back into a tree
+      def convert_run_list_trace(base)
+        @run_list_trace[base].map do |item|
+          skipped = @seen_items[item.type][item.name]
+          @seen_items[item.type][item.name] = true
```

Comment:
that feels a little bit dirty, since you have persistent state in the object and if you call this twice you'll 'remember' all the state from the previous call.  you probably want to have seen_items be a hash accumulator 2nd argument (default to {}) that you pass through to all the recursive child calls, then it'll be local state, and once you exit from the top level it'll go out of scope.